com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java

```diff
@@ -294,37 +110,49 @@ public Task<Void> create(IResource input) throws IOException, CompileExceptionEr
         Task.TaskBuilder<Void> taskBuilder = Task.<Void>newBuilder(this)
                 .setName(params.name())
                 .addInput(input)
-                .addOutput(input.changeExt(params.outExt()));
+                .addOutput(input.changeExt(params.outExt()))
+                .addOutput(input.changeExt(ComponentsCounter.EXT_COL));
         CollectionDesc.Builder builder = CollectionDesc.newBuilder();
         ProtoUtil.merge(input, builder);
-        Set<IResource> subCollections = new HashSet<IResource>();
+
+        Map<IResource, Integer> subCollections = new HashMap<>();
         collectSubCollections(builder, subCollections);
-        for (IResource subCollection : subCollections) {
+        for (IResource subCollection : subCollections.keySet()) {
             taskBuilder.addInput(subCollection);
+            IResource compCounterInput = input.getResource(ComponentsCounter.replaceExt(subCollection)).output();
+            taskBuilder.addInput(compCounterInput);
+            compCounterInputsCount.put(compCounterInput, subCollections.get(subCollection));
         }
 
         for (InstanceDesc inst : builder.getInstancesList()) {
             InstanceDesc.Builder instBuilder = InstanceDesc.newBuilder(inst);
             List<ComponentPropertyDesc> sourceProperties = instBuilder.getComponentPropertiesList();
             createResourcePropertyTasks(sourceProperties, input);
+            IResource res = project.getResource(inst.getPrototype());
+            IResource compCounterInput = input.getResource(ComponentsCounter.replaceExt(res)).output();
+            taskBuilder.addInputIfUnique(compCounterInput);
+            compCounterInputsCount.put(compCounterInput, compCounterInputsCount.getOrDefault(compCounterInput, 0) + 1);
         }
 
-        createGeneratedResources(this.project, builder);
-
+        Map<Long, IResource> uniqueResources = new HashMap<>();
+        createGeneratedResources(this.project, builder, uniqueResources);
+        
+        List<Task<?>> embedTasks = new ArrayList<>();
         for (long hash : uniqueResources.keySet()) {
             IResource genResource = uniqueResources.get(hash);
-
             taskBuilder.addOutput(genResource);
-
-            if (productsOfThisTask.contains(hash)) {
```

Comment:
`productsOfThisTask` is always equal to `uniqueResources` (see line 133-134)
I don't see the reson to have `productsOfThisTask`